gpbft/src/lib.rs

```diff
@@ -1,6 +1,184 @@
 // Copyright 2019-2024 ChainSafe Systems
 // SPDX-License-Identifier: Apache-2.0, MIT
 
+//! GPBFT consensus protocol implementation.
+//!
+//! This module provides the core structures and logic for the GPBFT consensus protocol, including:
+//! - Chain and tipset structures for blockchain state representation
+//! - Power table management for validator voting power tracking
+//! - Consensus phases and payload structures
+//! - Justification and verification mechanisms
+//! - Network types such as `ActorId`, `StoragePower`, and `PubKey`
+//!
+//! Key components:
+//! - [`ECChain`]: Represents a chain of tipsets
+//! - [`PowerTable`]: Manages validator voting power
+//! - [`Payload`]: Contains consensus round information
+//! - [`Justification`]: Holds votes and signatures for consensus decisions
+//!
+//! This module enables:
+//! - Building and validating blockchain structures
+//! - Managing validator power and keys
+//! - Progressing through consensus phases
+//! - Creating and verifying consensus decisions
+//!
+//! It provides the foundational types and logic for implementing GPBFT consensus in a blockchain network.
+
+pub mod api;
+pub mod chain;
+mod powertable;
 mod types;
 
-pub use types::PubKey;
+pub use powertable::{PowerEntries, PowerEntry, PowerTable};
+
+// re-exports
+pub use fvm_ipld_bitfield::BitField;
+pub use num_bigint::{BigInt, Sign};
+pub use num_traits::Zero;
+use strum_macros::Display;
+
+pub use crate::chain::{Cid, ECChain};
+pub use types::{ActorId, NetworkName, PubKey, StoragePower};
+
+/// Additional data signed by participants in a GPBFT instance
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct SupplementalData {
+    /// Merkle-tree root of instance-specific commitments
+    ///
+    /// Currently empty, but will eventually include things like
+    /// snark-friendly power-table commitments.
+    pub commitments: keccak_hash::H256,
+    /// The DagCBOR-blake2b256 CID of the power table used to validate the next instance
+    ///
+    /// This takes look-back into account and represents a `[]PowerEntry`.
+    /// The CID is limited to a maximum length of 38 bytes.
+    pub power_table: Cid,
+}
+
+/// Represents the different phases of the GPBFT consensus protocol
+#[repr(u8)]
+#[derive(Display, Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+#[strum(serialize_all = "UPPERCASE")]
+pub enum Phase {
```

Comment:
Phase is encoded as `uint8` in `go-f3` as part of `Payload`

Reply:
Hence the use of `repr(u8)`.